### ISSUE TITLE
Add device specific message flow debugging

### ISSUE THREAD

**Issue by pellmann** [eclipse/hono#319]
**Problem**

It is often hard to follow specific problems from end-to-end. This is true in particular with many different devices talking to Hono with different adapters on one side and also many different consumers and tenants on the other side.

It is hard for the solution developer to get an idea of the real message flow and what happens. It is also hard to identify a problem if a customers calls the DevOps with "My message never arrives..".

**Solution**

Allow an option in  the Device Registry to set a device (or a class of devices or a tenant) in a debug state. This could trigger different actions (dependent an the specific state):

- Extend the normal tracing/logging of the components, that are involved. Typically at least the Adapter and Hono Messaging.
- Add debugging information of each involved component to the message. This could be done with specified message properties.

WDYT?

**Comment by pellmann** [comment]
Dispatch Router uses x-opt-qd.trace (https://qpid.apache.org/releases/qpid-dispatch-0.8.0/book/book.html#amqp-mapping) as list of strings with all routers. So maybe something like x-opt-hono.trace might be nice.

**Comment by sophokles73** [comment]
I would rather employ existing technology like http://opentracing.io for this purpose. Maybe in conjunction with https://jaegertracing.io as the implementation.

**Comment by pellmann** [comment]
For overall traceability it would be great to have such a tracing technology as opentracing.

But I still see a value for tracing information in the message itself (e.g. in annotations).

When there will be a system with all components involved, and all of them use such a system and also each customer/tenant is able to see (only) his message traces it might substitute the in-message information. 

But I would assume it might be easier in a lot of installations to also have such a in-message tracing.

**Comment by sophokles73** [comment]
> But I still see a value for tracing information in the message itself (e.g. in annotations).

What kind of information would that be for example?

**Comment by pellmann** [comment]
- Component (adapter, messaging, router) and instance - to see the path, the message took
- Timestamp - to see the time consumed in each component/network part
- Configuration info - ssl in network and from device to adapter, ..
- System snapshot - open connections/cpu/ram

I think this all might be interesting in all variants of tracing. To put it in the message (if configured) is just one simple variant of bringing this information together and to scope it, because the message already has this 'trace'.

**Comment by sophokles73** [comment]
Well, I believe we can do all that with opentracing ...

**Comment by sophokles73** [comment]
fixed in https://github.com/eclipse/hono/commit/985fba492d37ca79aadcdfbaddb408f152ae68b5
